File: src/test/java/io/chesslave/model/PositionTest.java
package io.chesslave.model;

import javaslang.Tuple;
import javaslang.Tuple2;
import javaslang.collection.Set;
import javaslang.control.Option;
import org.junit.Before;
import org.junit.Test;

import java.util.NoSuchElementException;

import static org.hamcrest.CoreMatchers.hasItem;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

public class PositionTest {

    private Position startPosition;

    @Before
    public void setUp() {
        startPosition = Positions.fromText(
                "r|n|b|q|k|b|n|r",
                "p|p|p|p|p|p|p|p",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                "P|P|P|P|P|P|P|P",
                "R|N|B|Q|K|B|N|R");
    }

    @Test
    public void atTest() {
        final Option<Piece> whiteKing = startPosition.at(Square.of("e1"));
        assertTrue(whiteKing.isDefined());
        assertEquals(Piece.of(Piece.Type.KING, Color.WHITE), whiteKing.get());

        final Option<Piece> blackQueen = startPosition.at(Square.of("d8"));
        assertTrue(blackQueen.isDefined());
        assertEquals(Piece.of(Piece.Type.QUEEN, Color.BLACK), blackQueen.get());

        final Option<Piece> noPiece = startPosition.at(Square.of("b4"));
        assertTrue(noPiece.isEmpty());
    }

    @Test
    public void putTest() {
        final Position newPosition = startPosition.put(Square.of("e4"), Piece.of(Piece.Type.PAWN, Color.BLACK));
        final Option<Piece> whiteKing = newPosition.at(Square.of("e1"));
        assertTrue(whiteKing.isDefined());
        assertEquals(Piece.of(Piece.Type.KING, Color.WHITE), whiteKing.get());

        final Option<Piece> blackPawn = newPosition.at(Square.of("e4"));
        assertTrue(blackPawn.isDefined());
        assertEquals(Piece.of(Piece.Type.PAWN, Color.BLACK), blackPawn.get());

        final Option<Piece> noPiece = newPosition.at(Square.of("g3"));
        assertTrue(noPiece.isEmpty());
    }

    @Test
    public void removeTest() {
        final Position newPosition = startPosition.remove(Square.of("a2"));
        final Option<Piece> whiteKing = newPosition.at(Square.of("e1"));
        assertTrue(whiteKing.isDefined());
        assertEquals(Piece.of(Piece.Type.KING, Color.WHITE), whiteKing.get());

        final Option<Piece> noPiece = newPosition.at(Square.of("a2"));
        assertTrue(noPiece.isEmpty());
    }

    @Test
    public void moveTest() {
        final Position newPosition = startPosition.move(Square.of("e2"), Square.of("e4"));
        final Option<Piece> whiteKing = newPosition.at(Square.of("e1"));
        assertTrue(whiteKing.isDefined());
        assertEquals(Piece.of(Piece.Type.KING, Color.WHITE), whiteKing.get());

        final Option<Piece> whitePawn = newPosition.at(Square.of("e4"));
        assertTrue(whitePawn.isDefined());
        assertEquals(Piece.of(Piece.Type.PAWN, Color.WHITE), whitePawn.get());

        final Option<Piece> noPiece = newPosition.at(Square.of("e2"));
        assertTrue(noPiece.isEmpty());
    }

    @Test
    public void moveToCaptureTest() {
        final Position newPosition = startPosition.move(Square.of("e2"), Square.of("e7"));
        final Option<Piece> whitePawn = newPosition.at(Square.of("e7"));
        assertTrue(whitePawn.isDefined());
        assertEquals(Piece.of(Piece.Type.PAWN, Color.WHITE), whitePawn.get());

        final Option<Piece> noPiece = newPosition.at(Square.of("e2"));
        assertTrue(noPiece.isEmpty());
    }

    @Test(expected = NoSuchElementException.class)
    public void moveEmptySquareTest() {
        startPosition.move(Square.of("e4"), Square.of("e2"));
    }

    @Test
    public void toSetTest() {
        Set<Tuple2<Square, Piece>> positionSet = startPosition.toSet();
        assertEquals(32, positionSet.size());
        assertThat(positionSet, hasItem(Tuple.of(Square.of("a1"), Piece.of(Piece.Type.ROOK, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("b1"), Piece.of(Piece.Type.KNIGHT, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("c1"), Piece.of(Piece.Type.BISHOP, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("d1"), Piece.of(Piece.Type.QUEEN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("e1"), Piece.of(Piece.Type.KING, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("f1"), Piece.of(Piece.Type.BISHOP, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("g1"), Piece.of(Piece.Type.KNIGHT, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("h1"), Piece.of(Piece.Type.ROOK, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("a2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("b2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("c2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("d2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("e2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("f2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("g2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("h2"), Piece.of(Piece.Type.PAWN, Color.WHITE))));

        assertThat(positionSet, hasItem(Tuple.of(Square.of("a8"), Piece.of(Piece.Type.ROOK, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("b8"), Piece.of(Piece.Type.KNIGHT, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("c8"), Piece.of(Piece.Type.BISHOP, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("d8"), Piece.of(Piece.Type.QUEEN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("e8"), Piece.of(Piece.Type.KING, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("f8"), Piece.of(Piece.Type.BISHOP, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("g8"), Piece.of(Piece.Type.KNIGHT, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("h8"), Piece.of(Piece.Type.ROOK, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("a7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("b7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("c7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("d7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("e7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("f7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("g7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
        assertThat(positionSet, hasItem(Tuple.of(Square.of("h7"), Piece.of(Piece.Type.PAWN, Color.BLACK))));
    }

    @Test
    public void equalsTest() {
        assertTrue(startPosition.equals(startPosition));
        assertFalse(startPosition.equals(new Object()));
        final Position samePosition = Positions.fromText(
                "r|n|b|q|k|b|n|r",
                "p|p|p|p|p|p|p|p",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                "P|P|P|P|P|P|P|P",
                "R|N|B|Q|K|B|N|R");
        assertTrue(startPosition.equals(samePosition));
    }

    @Test
    public void hashCodeTest() {
        final int startPositionHash = startPosition.hashCode();
        assertEquals(startPositionHash, startPosition.hashCode());
        assertNotEquals(startPositionHash, new Object().hashCode());
        final Position samePosition = Positions.fromText(
                "r|n|b|q|k|b|n|r",
                "p|p|p|p|p|p|p|p",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                " | | | | | | | ",
                "P|P|P|P|P|P|P|P",
                "R|N|B|Q|K|B|N|R");
        assertEquals(startPositionHash, samePosition.hashCode());
    }

    @Test
    public void toStringTest() {
        final String expected =
                "r|n|b|q|k|b|n|r\n" +
                "p|p|p|p|p|p|p|p\n" +
                " | | | | | | | \n" +
                " | | | | | | | \n" +
                " | | | | | | | \n" +
                " | | | | | | | \n" +
                "P|P|P|P|P|P|P|P\n" +
                "R|N|B|Q|K|B|N|R";
        assertEquals(expected, startPosition.toString());
    }
}

File: backend/src/test/java/io/chesslave/model/PositionTest.kt
package io.chesslave.model

import io.chesslave.model.Piece.Type
import io.vavr.Tuple
import org.hamcrest.CoreMatchers.hasItem
import org.junit.Assert.*
import org.junit.Test
import java.util.*

internal val initialPosition = Position.Builder()
    .withPiece(Board.a1, Piece.whiteRook)
    .withPiece(Board.b1, Piece.whiteKnight)
    .withPiece(Board.c1, Piece.whiteBishop)
    .withPiece(Board.d1, Piece.whiteQueen)
    .withPiece(Board.e1, Piece.whiteKing)
    .withPiece(Board.f1, Piece.whiteBishop)
    .withPiece(Board.g1, Piece.whiteKnight)
    .withPiece(Board.h1, Piece.whiteRook)
    .withPiece(Board.a2, Piece.whitePawn)
    .withPiece(Board.b2, Piece.whitePawn)
    .withPiece(Board.c2, Piece.whitePawn)
    .withPiece(Board.d2, Piece.whitePawn)
    .withPiece(Board.e2, Piece.whitePawn)
    .withPiece(Board.f2, Piece.whitePawn)
    .withPiece(Board.g2, Piece.whitePawn)
    .withPiece(Board.h2, Piece.whitePawn)
    .withPiece(Board.a8, Piece.blackRook)
    .withPiece(Board.b8, Piece.blackKnight)
    .withPiece(Board.c8, Piece.blackBishop)
    .withPiece(Board.d8, Piece.blackQueen)
    .withPiece(Board.e8, Piece.blackKing)
    .withPiece(Board.f8, Piece.blackBishop)
    .withPiece(Board.g8, Piece.blackKnight)
    .withPiece(Board.h8, Piece.blackRook)
    .withPiece(Board.a7, Piece.blackPawn)
    .withPiece(Board.b7, Piece.blackPawn)
    .withPiece(Board.c7, Piece.blackPawn)
    .withPiece(Board.d7, Piece.blackPawn)
    .withPiece(Board.e7, Piece.blackPawn)
    .withPiece(Board.f7, Piece.blackPawn)
    .withPiece(Board.g7, Piece.blackPawn)
    .withPiece(Board.h7, Piece.blackPawn)
    .build()

class PositionTest {

    @Test
    fun atTest() {
        val whiteKing = initialPosition.at(Board.e1)
        assertTrue(whiteKing != null)
        assertEquals(Piece.whiteKing, whiteKing)

        val blackQueen = initialPosition.at(Board.d8)
        assertTrue(blackQueen != null)
        assertEquals(Piece.blackQueen, blackQueen)

        val noPiece = initialPosition.at(Board.b4)
        assertTrue(noPiece == null)
    }

    @Test
    fun putTest() {
        val newPosition = initialPosition.put(Board.e4, Piece.blackPawn)
        val whiteKing = newPosition.at(Board.e1)
        assertTrue(whiteKing != null)
        assertEquals(Piece.whiteKing, whiteKing)

        val blackPawn = newPosition.at(Board.e4)
        assertTrue(blackPawn != null)
        assertEquals(Piece.blackPawn, blackPawn)

        val noPiece = newPosition.at(Board.g3)
        assertTrue(noPiece == null)
    }

    @Test
    fun removeTest() {
        val newPosition = initialPosition.remove(Board.a2)
        val whiteKing = newPosition.at(Board.e1)
        assertTrue(whiteKing != null)
        assertEquals(Piece.whiteKing, whiteKing)

        val noPiece = newPosition.at(Board.a2)
        assertTrue(noPiece == null)
    }

    @Test
    fun moveTest() {
        val newPosition = initialPosition.move(Board.e2, Board.e4)
        val whiteKing = newPosition.at(Board.e1)
        assertTrue(whiteKing != null)
        assertEquals(Piece.whiteKing, whiteKing)

        val whitePawn = newPosition.at(Board.e4)
        assertTrue(whitePawn != null)
        assertEquals(Piece.whitePawn, whitePawn)

        val noPiece = newPosition.at(Board.e2)
        assertTrue(noPiece == null)
    }

    @Test
    fun moveToCaptureTest() {
        val newPosition = initialPosition.move(Board.e2, Board.e7)
        val whitePawn = newPosition.at(Board.e7)
        assertTrue(whitePawn != null)
        assertEquals(Piece.whitePawn, whitePawn)

        val noPiece = newPosition.at(Board.e2)
        assertTrue(noPiece == null)
    }

    @Test(expected = NoSuchElementException::class)
    fun moveEmptySquareTest() {
        initialPosition.move(Board.e4, Board.e2)
    }

    @Test
    fun toSetTest() {
        val positionSet = initialPosition.toSet()
        assertEquals(32, positionSet.size().toLong())
        assertThat(positionSet, hasItem(Tuple.of(Board.a1, Piece.whiteRook)))
        assertThat(positionSet, hasItem(Tuple.of(Board.b1, Piece.whiteKnight)))
        assertThat(positionSet, hasItem(Tuple.of(Board.c1, Piece.whiteBishop)))
        assertThat(positionSet, hasItem(Tuple.of(Board.d1, Piece.whiteQueen)))
        assertThat(positionSet, hasItem(Tuple.of(Board.e1, Piece.whiteKing)))
        assertThat(positionSet, hasItem(Tuple.of(Board.f1, Piece.whiteBishop)))
        assertThat(positionSet, hasItem(Tuple.of(Board.g1, Piece.whiteKnight)))
        assertThat(positionSet, hasItem(Tuple.of(Board.h1, Piece.whiteRook)))
        assertThat(positionSet, hasItem(Tuple.of(Board.a2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.b2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.c2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.d2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.e2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.f2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.g2, Piece.whitePawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.h2, Piece.whitePawn)))

        assertThat(positionSet, hasItem(Tuple.of(Board.a8, Piece.blackRook)))
        assertThat(positionSet, hasItem(Tuple.of(Board.b8, Piece.blackKnight)))
        assertThat(positionSet, hasItem(Tuple.of(Board.c8, Piece.blackBishop)))
        assertThat(positionSet, hasItem(Tuple.of(Board.d8, Piece.blackQueen)))
        assertThat(positionSet, hasItem(Tuple.of(Board.e8, Piece.blackKing)))
        assertThat(positionSet, hasItem(Tuple.of(Board.f8, Piece.blackBishop)))
        assertThat(positionSet, hasItem(Tuple.of(Board.g8, Piece.blackKnight)))
        assertThat(positionSet, hasItem(Tuple.of(Board.h8, Piece.blackRook)))
        assertThat(positionSet, hasItem(Tuple.of(Board.a7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.b7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.c7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.d7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.e7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.f7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.g7, Piece.blackPawn)))
        assertThat(positionSet, hasItem(Tuple.of(Board.h7, Piece.blackPawn)))
    }

    @Test
    fun equalsTest() {
        assertEquals(initialPosition, initialPosition)
        assertNotEquals(initialPosition, Any())
        val samePosition = positionFromText(
            "r|n|b|q|k|b|n|r",
            "p|p|p|p|p|p|p|p",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            "P|P|P|P|P|P|P|P",
            "R|N|B|Q|K|B|N|R")
        assertEquals(initialPosition, samePosition)
    }

    @Test
    fun hashCodeTest() {
        val startPositionHash = initialPosition.hashCode()
        assertEquals(startPositionHash, initialPosition.hashCode())
        assertNotEquals(startPositionHash, Any().hashCode())
        val samePosition = positionFromText(
            "r|n|b|q|k|b|n|r",
            "p|p|p|p|p|p|p|p",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            "P|P|P|P|P|P|P|P",
            "R|N|B|Q|K|B|N|R")
        assertEquals(startPositionHash, samePosition.hashCode())
    }

    @Test
    fun toStringTest() {
        val expected = "" +
            "r|n|b|q|k|b|n|r\n" +
            "p|p|p|p|p|p|p|p\n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            "P|P|P|P|P|P|P|P\n" +
            "R|N|B|Q|K|B|N|R"
        assertEquals(expected, initialPosition.toString())
    }
}

class positionFromTextTest {

    @Test
    fun canCreatePositionFromText() {
        val got = positionFromText(
            "r|n|b|q|k|b|n|r",
            "p|p|p|p|p|p|p|p",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            " | | | | | | | ",
            "P|P|P|P|P|P|P|P",
            "R|N|B|Q|K|B|N|R")
        assertEquals(initialPosition, got)
    }
}

class positionToTextTest {

    @Test
    fun canRenderPositionToText() {
        val got = positionToText(initialPosition)
        val expected = "" +
            "r|n|b|q|k|b|n|r\n" +
            "p|p|p|p|p|p|p|p\n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            " | | | | | | | \n" +
            "P|P|P|P|P|P|P|P\n" +
            "R|N|B|Q|K|B|N|R"
        assertEquals(expected, got)
    }
}